### PR TITLE
Fall back on shared dictionaries when available

### PR DESCRIPTION
**Pull request by HuamengJiang**
When flatmap subcolumns have shared dictionaries across keys, they will be under the reserved sequence number 0.
This PR introduces logic to fall back on sequence 0 shared dictionary data streams when dictionary data stream for the current sequence is missing.

Test plan - added unit tests for dictionary provider and extended existing file based unit tests for flatmap.

````````````````````````
== NO RELEASE NOTE ==
`````````````````````````

### CODE REVIEW

**Comment by sdruzkin** [comment]
Add the to the PR description.

```
== NO RELEASE NOTE ==
```

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java
Comment (sdruzkin) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java:96]: Please don't inline requireNonNull into the if-statement, it significantly reduces readability.

Do we care about the returned dictionaryDataStream.openStream?

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java
Comment (sdruzkin) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java:106]: Extract `loadDictionary(....` into a separate variable to improve readability.

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java
Comment (sdruzkin) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java:110]: checkArgument can work with the %s string formats, replace this with 
```
checkArgument(entry.size == items, "Shared dictionary size mismatch for stream: %s", streamDescriptor);
```

File: presto-orc/src/main/java/com/facebook/presto/orc/StreamDescriptor.java
Comment (sdruzkin) [inline presto-orc/src/main/java/com/facebook/presto/orc/StreamDescriptor.java:52]: Give it a better name, or at least don't shorten the words -> `dup`.

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java
Comment (sdruzkin) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java:101]: See if you can use `sharedDictionaries.computeIfAbsent(...)` instead of `isNew`

File: presto-orc/src/main/java/com/facebook/presto/orc/StripeReader.java
Comment (HuamengJiang) [reply presto-orc/src/main/java/com/facebook/presto/orc/StripeReader.java:308]: That's gonna be a larger better engineering change. I'll file a task to track instead.

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java
Comment (HuamengJiang) [reply presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java:96]: We do. In fact, that's the gist of the fallback mechanism. Curious actually, if we don't need to inline, why do we have the signature for requireNonNull? Why not just have a dedicated check not null statement?

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java
Comment (HuamengJiang) [reply presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java:101]: Do you mean setting the boolean with the input functions to computeIfAbsent? The goal here really is to know whether insertion happens, just for the sake of the return value.

Comment (HuamengJiang) [reply presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java:101]: Yeah I think the function given to computeIfAbsent is a remapping function. The main goal of isNew here is to record whether an insertion happened or not. Maybe the way I'm writing it is not very idiomatic and I should just use containsKey and live with the additional lookup?

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java
Comment (sdruzkin) [reply presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java:96]: > why do we have the signature for requireNonNull?
Different teams have different requirements, for example in this case I try to avoid two levels of nesting. 
Having requireNonNull return the input argument is very when assigning fields in constructors.

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java
Comment (sdruzkin) [reply presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java:101]: computeIfAbsent is only invoked when the key is absent, imho it should work nice in this case. I'm leaving the final decision to use it or not up to you, it's more like a personal preference.

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java
Comment (arunthirupathi) [reply presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java:58]: If this is only used within the file, make it private static.

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java
Comment (arunthirupathi) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java:97]: I see 2 logic changes in the new code.

1. Addition of the condiition dictionaryDataStream.openStream() != null.
2. requireNonNull on the return value.

If the logic change has to be minimized. 

I would propose the following code to minimize the combinations. This ensures that when the stream is directly found, it uses it. 

if (dictionaryDataStream.openStream() != null) {
 return loadDictionary.
}

checkArgument( sequence != DEFAULT_SEQUENCE_ID);

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java
Comment (arunthirupathi) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java:106]: requireNonNull for sharedDictionaryDataStream.

File: presto-orc/src/test/java/com/facebook/presto/orc/TestLongDictionaryProvider.java
Comment (arunthirupathi) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestLongDictionaryProvider.java:377]: Java equals require to handle the Object o being null and the Object not an instance of NodeId. 

Generally the second case after o == this is

if (!o instanceof NodeId) return false;

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java
Comment (HuamengJiang) [reply presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java:66]: It could be. I'm slightly concerned about the amount of double checking we are doing. DictionaryBuffer resizing logic is handled inside the class and tested, and the long input stream logic should also be well tested. I think they should have a little bit of branching cost right?

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java
Comment (HuamengJiang) [reply presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java:58]: I'll name it to SharedDictionaryContainer and make it private static.

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java
Comment (HuamengJiang) [reply presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java:96]: I'm thinking adding and naming the constant ColumnEncoding.RESERVED_SEQUENCE_ID or something more semantic. WDYT?

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java
Comment (HuamengJiang) [reply presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java:97]: That's fair enough. I was keeping this branching logic as similar to bbio as possible, though there's no production code that relies on bypassing shared dictionary lookup through sequence 0.

File: presto-orc/src/test/java/com/facebook/presto/orc/TestLongDictionaryProvider.java
Comment (HuamengJiang) [reply presto-orc/src/test/java/com/facebook/presto/orc/TestLongDictionaryProvider.java:377]: My bad, I thought of this earlier but forgot to include it.

File: presto-orc/src/main/java/com/facebook/presto/orc/StripeReader.java
Comment (HuamengJiang) [reply presto-orc/src/main/java/com/facebook/presto/orc/StripeReader.java:305]: Yeah the streams are handled correctly. Will remove this comment.

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java
Comment (HuamengJiang) [reply presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java:106]: I could  add it, I'm slightly concerned because we already know this is never null, but it could be missingSource which returns null when opened, so the latter check is more important.

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java
Comment (arunthirupathi) [reply presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java]: I would keep it just as DEFAULT_SEQUENCE_ID

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java
Comment (arunthirupathi) [reply presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java:67]: Since this is a private class, if we are not checking this is ok.

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java
Comment (arunthirupathi) [reply presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java]: Call it just SharedDictionary

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java
Comment (arunthirupathi) [reply presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java:109]: I am thinking of the case, where there is a shared dictionary in flat map but the shared dictionary stream is missing in the file (due to writer bugs or file corruption)

Comment (HuamengJiang) [reply presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java:109]: What I meant is that when the shared dictionary stream is indeed missing, the return value is a `missingStreamSource(streamType)`, which is only null when calling `openStream()`. We do have the check for when the stream is opened.

Comment (arunthirupathi) [reply presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java:109]: I see makes sense, then, no check needs to be added.

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java
Comment (arunthirupathi) [reply presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java]: Can we change the code to the following logic.

if (dictionaryDataStream.openStream() != null) {
return loadDictionary.
}

checkArgument( sequence != DEFAULT_SEQUENCE_ID);

**Review by sdruzkin — APPROVED** [review]
LGTM after addressing Arun's notes.

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java
Comment (arunthirupathi) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java:95]: why do we have requireNonNull here ? From your previous explanation, we always receive MissingStreamSource here.

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java
Comment (arunthirupathi) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java:98]: As suggested, can we change the logic here.

Comment (HuamengJiang) [reply presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java:98]: It's good that we double-checked here. I realized that we can completely delete the default sequence id check and keep the same behavior and error handling. The same can be done for bbio as well.

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java
Comment (arunthirupathi) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java:68]: Can you undo this comment formatting?

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java
Comment (arunthirupathi) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java:103]: Can we add a checkState(streamDescriptor.sequence != DEFAULT_SEQUENCE_ID) ?

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java
Comment (arunthirupathi) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java:114]: use checkState instead of checkArgument. checkArgument implies one of the parameter passed in is incorrect. checkState implies an internal constraint is violated.

File: presto-orc/src/test/java/com/facebook/presto/orc/TestLongDictionaryProvider.java
Comment (arunthirupathi) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestLongDictionaryProvider.java:390]: Can we name this class differently ? Case implies a Test case. If there is no better name, we can name it as SharedDictionaryTestInput

**Review by arunthirupathi — APPROVED** [review]
Looks good, once the comments are addressed.

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java
Comment (HuamengJiang) [reply presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java:103]: The reason it's not here is that we want to handle the corruptions in one place. We could additionally throw here, but it's a slightly less optimal code organization.

File: presto-orc/src/test/java/com/facebook/presto/orc/TestLongDictionaryProvider.java
Comment (HuamengJiang) [reply presto-orc/src/test/java/com/facebook/presto/orc/TestLongDictionaryProvider.java:390]: It is a test case though? It's a struct of inputs and expected states. Test input also sounds good to me.

**Comment by arunthirupathi** [comment]
This has 2 commits now, can you please cleanup and push only one commit ?

**Review by arunthirupathi — APPROVED** [review]
Looks good to me.

**Comment by highker** [comment]
Quick nits: let's lower case the commit title: "Fall back on shared dictionaries when available"

File: presto-orc/src/main/java/com/facebook/presto/orc/StreamDescriptor.java
Comment (highker) [inline presto-orc/src/main/java/com/facebook/presto/orc/StreamDescriptor.java:54]: remove all `this.`:
```java
return new StreamDescriptor(streamName, streamId, fieldName, orcType, orcDataSource, nestedStreams, sequence);

**Review by highker — APPROVED** [review]

**Comment by highker** [comment]
Could you rebase and force push. The tests failure might be related due to lack of recent changes.

**Pull request merged**
